jax/_src/pallas/primitives.py

```diff
@@ -993,3 +993,267 @@ def _lower_fun(*lower_fun_args):
     return out[:num_return_values]
 
   return mlir.lower_fun(_lower_fun, multiple_results=True)(ctx, *args)
+
+
+def _get_ref_and_transforms(ref):
+  if isinstance(ref, state.TransformedRef):
+    return ref.ref, ref.transforms
+  return ref, ()
+
+
+class DeviceIdType(enum.Enum):
+  MESH = "mesh"
+  LOGICAL = "logical"
+
+
+def check_sem_avals(
+    sem_aval, sem_transforms_avals, name, allowed_semaphore_types=None
+):
+  if allowed_semaphore_types is None:
+    allowed_semaphore_types = {
+        pallas_core.semaphore,
+        pallas_core.barrier_semaphore,
+        # For interpret mode.
+        pallas_core.SEMAPHORE_INTERPRET_DTYPE,
+    }
+  if not isinstance(sem_aval, state.AbstractRef):
+    raise ValueError(f"Cannot {name} on a non-semaphore Ref: {sem_aval}")
+  sem_shape = sem_aval.shape
+  if sem_transforms_avals:
+    sem_shape = sem_transforms_avals[-1].get_indexer_shape()
+  if sem_shape:
+    raise ValueError(f"Cannot {name} on a non-()-shaped semaphore: {sem_shape}")
+  # Uncomment when semaphore type works for Mosaic-GPU lowering
+  # sem_dtype = sem_aval.dtype
+  # if not any(
+  #     jnp.issubdtype(sem_dtype, sem_type)
+  #     for sem_type in allowed_semaphore_types
+  # ):
+  #   raise ValueError(
+  #       f"Must {name} semaphores of the following types:"
+  #       f" {allowed_semaphore_types}."
+  #   )
```

Comment:
Please uncomment this for now. As discussed, I'll figure out the bits to make this work with the Mosaic GPU lowering